src/Build/Graph/ProjectGraph.cs

```diff
@@ -22,21 +26,56 @@ internal sealed class ProjectGraph
         /// <exception cref="InvalidProjectFileException">If the evaluation of any project in the graph fails.</exception>
         public ProjectGraph(string entryProjectFile)
         {
-            var graphNode =
-                new ProjectGraphNode(ProjectCollection.GlobalProjectCollection.LoadProject(entryProjectFile));
-            _projectNodes.Add(graphNode);
+            LoadGraph(entryProjectFile);
         }
 
         /// <summary>
-        /// Constructs a graph starting from the given project files.
+        /// Get an unordered collection of all project nodes in the graph.
         /// </summary>
-        /// <param name="entryProjectFiles">The project files to use as the entry points in constructing the graph</param>
-        /// <exception cref="InvalidProjectFileException">If the evaluation of any project in the graph fails.</exception>
-        public ProjectGraph(IEnumerable<string> entryProjectFiles) { }
+        public IReadOnlyCollection<ProjectGraphNode> ProjectNodes => _allParsedProjects.Values;
+
+        private ProjectGraphNode CreateNewNode(string projectFilePath)
+        {
+            var graphNode =
+                new ProjectGraphNode(ProjectCollection.GlobalProjectCollection.LoadProject(projectFilePath));
+            _allParsedProjects.Add(projectFilePath, graphNode);
+            return graphNode;
+        }
 
         /// <summary>
-        /// Get an unordered collection of all project nodes in the graph.
+        /// Load a graph with root node at entryProjectFile
+        /// Maintain a queue of projects to be processed- each queue item is a key value pair of the project to be evaluated and its parent
+        /// Once the project has been evaluated, add a project reference to this evaluated target from the parent node
         /// </summary>
-        public IReadOnlyCollection<ProjectGraphNode> ProjectNodes => _projectNodes;
+        /// <param name="entryProjectFile"></param>
+        private void LoadGraph(string entryProjectFile)
+        {
+            var projectsToEvaluate = new Queue<KeyValuePair<string, ProjectGraphNode>>();
```

Comment:
This might read a bit more naturally if these were tuples rather than `KeyValuePair` but I don't feel strongly about it.